docs/assets/js/search.js

```diff
@@ -0,0 +1,102 @@
+import Fuse from "https://cdn.jsdelivr.net/npm/fuse.js@7.0.0/dist/fuse.mjs";
+import { q } from "./util.js";
+import { resetToCategory } from "./nodes.js";
+
+const getFuse = (() => {
+  let fuse;
+  return () => {
+    if (fuse) return fuse;
+    fuse = getSearchIndex().then(
+      (searchIndex) =>
+        new Fuse(searchIndex, {
+          includeMatches: true,
+          minMatchCharLength: 2,
+          keys: [
+            { name: "title", score: 0.5 },
+            { name: "category.name", score: 1 },
+            { name: "category.id", score: 0.8 },
+          ],
+        })
+    );
+    return fuse;
+  };
+})();
+
+const getSearchIndex = (() => {
+  let searchIndex;
+  return () => {
+    if (searchIndex) return searchIndex;
+    searchIndex = fetch("/assets/js/search_index.json").then((res) =>
+      res.json()
+    );
+    return searchIndex;
+  };
+})();
+
+const searchDebounceMs = 200;
+
+export const setupSearch = async () => {
+  // wait for fuse/searchIndex to load before setting up listeners
+  await getFuse();
```

Comment:
we probably should look into a nicer loading state but for now I changed the placeholder on load to "Loading" and then here we wait for fuse to be ready (+ search index internally), before updating the placeholder and adding listeners.